src/backend/distributed/planner/function_call_delegation.c

```diff
@@ -384,3 +351,95 @@ TryToDelegateFunctionCall(DistributedPlanningContext *planContext)
 
 	return FinalizePlan(planContext->plan, distributedPlan);
 }
+
+
+/*
+ * ShardPlacementWhenColocatedWithDistTable decides on a placement
+ * for delegating a procedure call that accesses a distributed table.
+ */
+static ShardPlacement *
+ShardPlacementWhenColocatedWithDistTable(DistObjectCacheEntry *procedure,
+										 FuncExpr *funcExpr,
+										 Var *partitionColumn,
+										 CitusTableCacheEntry *cacheEntry,
+										 PlannedStmt *plan)
+{
+	if (procedure->distributionArgIndex < 0 ||
+		procedure->distributionArgIndex >= list_length(funcExpr->args))
+	{
+		ereport(DEBUG1, (errmsg("function call does not have a distribution argument")));
+		return NULL;
+	}
+
+	Const *partitionValue = (Const *) list_nth(funcExpr->args,
+											   procedure->distributionArgIndex);
+
+	if (IsA(partitionValue, Param))
+	{
+		Param *partitionParam = (Param *) partitionValue;
+
+		if (partitionParam->paramkind == PARAM_EXTERN)
+		{
+			/* Don't log a message, we should end up here again without a parameter */
+			DissuadePlannerFromUsingPlan(plan);
+			return NULL;
+		}
+	}
+
+	if (!IsA(partitionValue, Const))
+	{
+		ereport(DEBUG1, (errmsg("distribution argument value must be a constant")));
+		return NULL;
+	}
+
+	Datum partitionValueDatum = partitionValue->constvalue;
+
+	if (partitionValue->consttype != partitionColumn->vartype)
+	{
+		bool missingOk = false;
+		partitionValue =
+			TransformPartitionRestrictionValue(partitionColumn, partitionValue,
+											   missingOk);
+	}
+
+	ShardInterval *shardInterval = FindShardInterval(partitionValueDatum, cacheEntry);
+	if (shardInterval == NULL)
+	{
+		ereport(DEBUG1, (errmsg("cannot push down call, failed to find shard interval")));
+		return NULL;
+	}
+
+	List *placementList = ActiveShardPlacementList(shardInterval->shardId);
+	if (list_length(placementList) != 1)
+	{
+		/* punt on this for now */
+		ereport(DEBUG1, (errmsg(
+							 "cannot push down function call for replicated distributed tables")));
+		return NULL;
+	}
+
+	return linitial(placementList);
+}
+
+
+/*
+ * ShardPlacementWhenColocatedWithReferenceTable decides on a placement for delegating
+ * a procedure call that reads from a reference table.
+ *
+ * If citus.task_assignment_policy is set to round-robin, we assign a different placement
+ * on consecutive runs. Otherwise the function returns the first placement available.
+ */
+static ShardPlacement *
+ShardPlacementWhenColocatedWithReferenceTable(CitusTableCacheEntry *cacheEntry)
+{
+	const ShardInterval *shardInterval = cacheEntry->sortedShardIntervalArray[0];
+	const uint64 referenceTableShardId = shardInterval->shardId;
+	List *placementList = ActiveShardPlacementList(referenceTableShardId);
+
+	if (TaskAssignmentPolicy == TASK_ASSIGNMENT_ROUND_ROBIN)
+	{
+		placementList = RoundRobinReorder(NULL, placementList);
```

Comment:
I think the first argument is never used. We should maybe just remove it from the reorderFunction signature.